python/ray/tune/impl/tuner_internal.py

```diff
@@ -0,0 +1,214 @@
+            raise TuneError("You need to provide a trainable to tune.")
+
+        self._is_restored = False
+        self._trainable = trainable
+        self._tune_config = tune_config
+        self._run_config = run_config
+        self._experiment_checkpoint_dir = self._get_or_create_experiment_checkpoint_dir(
+            self._run_config
+        )
+
+        # Not used for restored Tuner.
+        self._param_space = param_space
+        self._process_dataset_param()
+
+        # This needs to happen before `tune.run()` is kicked in.
+        # This is because currently tune does not exit gracefully if
+        # run in ray client mode - if crash happens, it just exits immediately
+        # without allowing for checkpointing tuner and trainable.
+        # Thus this has to happen before tune.run() so that we can have something
+        # to restore from.
+        tuner_ckpt = os.path.join(self._experiment_checkpoint_dir, _TUNER_PKL)
+        with open(tuner_ckpt, "wb") as fp:
+            pickle.dump(self, fp)
+
+        trainable_ckpt = os.path.join(self._experiment_checkpoint_dir, _TRAINABLE_PKL)
+        with open(trainable_ckpt, "wb") as fp:
+            pickle.dump(self._trainable, fp)
+
+    def _process_dataset_param(self):
+        """Dataset needs to be fully executed before sent over to trainables.
+
+        A valid dataset configuration in param space looks like:
+        "datasets": {
+            "train_dataset": tune.grid_search([ds1, ds2]),
+        },
+        """
+
+        def _helper(dataset_dict: dict):
+            for k, v in dataset_dict.items():
+                if isinstance(v, dict):
+                    _helper(v)
+                elif isinstance(v, Dataset):
+                    dataset_dict[k] = v.fully_executed()
+                # TODO(xwjiang): Consider CV config for beta.
+                # elif isinstance(v, int):
+                #     # CV settings
+                #     pass
+                elif isinstance(v, list):
+                    if not all([isinstance(v_item, int) for v_item in v]) and not all(
+                        [isinstance(v_item, Dataset) for v_item in v]
+                    ):
+                        raise TuneError(
+                            "Wrongly formatted dataset param passed in Tune!"
+                        )
+                    if len(v) > 0 and isinstance(v[0], Dataset):
+                        dataset_dict[k] = [v_item.fully_executed() for v_item in v]
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        for key in _DATASET_KEYS:
+            if key in self._param_space:
+                ds = self._param_space[key]
+                if isinstance(ds, Dataset):
+                    self._param_space[key] = ds.fully_executed()
+                elif isinstance(ds, dict):
+                    _helper(ds)
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
```

Comment:
Can we just check all config keys?

If I see this correctly it only works with `tune.grid_search` (because it will be converted into a dict) but not with `tune.choice()` (as it will be a `Categorical` object). Can we also support `tune.choice`?